File: app/src/main/java/com/example/mike/mp3player/client/view/fragments/PlaybackTrackerFragment.java
package com.example.mike.mp3player.client.view.fragments;

import android.os.Bundle;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.mike.mp3player.R;
import com.example.mike.mp3player.client.utils.TimerUtils;
import com.example.mike.mp3player.client.view.SeekerBar;
import com.example.mike.mp3player.client.view.TimeCounter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public class PlaybackTrackerFragment extends Fragment {

    private TextView duration;
    private SeekerBar seekerBar;
    private TimeCounter counter;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.fragment_playback_tracker, container, true);
    }

    @Override
    public void onViewCreated(View view, Bundle bundle) {
        super.onViewCreated(view, bundle);
        TextView counterView = view.findViewById(R.id.timer);
        this.counter = new TimeCounter( counterView);

        this.setSeekerBar(view.findViewById(R.id.seekBar));
        this.getSeekerBar().init();
        this.getSeekerBar().setTimeCounter(counter);
        this.duration = view.findViewById(R.id.duration);
    }

    public TimeCounter getCounter() {
        return counter;
    }

    public void setMetaData(MediaMetadataCompat metaData) {
        long duration = metaData.getLong(MediaMetadataCompat.METADATA_KEY_DURATION);
        getCounter().setDuration(duration);
        this.duration.setText(TimerUtils.formatTime(duration));
        getSeekerBar().getMySeekerMediaControllerCallback().onMetadataChanged(metaData);
    }

    public void setPlaybackState(PlaybackStateCompat state) {
        getCounter().updateState(state);
        getSeekerBar().getMySeekerMediaControllerCallback().onPlaybackStateChanged(state);
    }

    public SeekerBar getSeekerBar() {
        return seekerBar;
    }

    public void setSeekerBar(SeekerBar seekerBar) {
        this.seekerBar = seekerBar;
    }
}

File: client/src/main/java/com/github/goldy1992/mp3player/client/views/fragments/PlaybackTrackerFragment.kt
package com.github.goldy1992.mp3player.client.views.fragments

import android.os.Bundle
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import com.github.goldy1992.mp3player.client.MediaControllerAdapter
import com.github.goldy1992.mp3player.client.SeekerBarController2
import com.github.goldy1992.mp3player.client.databinding.FragmentPlaybackTrackerBinding
import com.github.goldy1992.mp3player.client.utils.TimerUtils.formatTime
import com.github.goldy1992.mp3player.client.views.TimeCounter
import com.github.goldy1992.mp3player.commons.LogTagger
import com.google.android.material.slider.Slider
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class PlaybackTrackerFragment : Fragment(), LogTagger, Observer<Any> {

    @Inject
    lateinit var mediaControllerAdapter : MediaControllerAdapter

    @Inject
    lateinit var seekerBarController: SeekerBarController2

    @Inject
    lateinit var counter: TimeCounter

    private var duration: TextView? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val binding = FragmentPlaybackTrackerBinding.inflate(inflater)
        val counterView = binding.timer
        counter.init(counterView)
        val seekerBar: Slider = binding.seekBar
        seekerBarController.init(seekerBar)
        duration = binding.duration
        mediaControllerAdapter.playbackState.observe(viewLifecycleOwner, seekerBarController)
        mediaControllerAdapter.metadata.observe(viewLifecycleOwner, seekerBarController)
        mediaControllerAdapter.repeatMode.observe(viewLifecycleOwner, seekerBarController)
        mediaControllerAdapter.playbackState.observe(viewLifecycleOwner, this)
        mediaControllerAdapter.metadata.observe(viewLifecycleOwner, this)
        return binding.root
    }

    override fun logTag(): String {
        return "PLY_TRKR_FRGMT"
    }

    override fun onChanged(data : Any) {
        if (data is MediaMetadataCompat) {
            val duration = data.getLong(MediaMetadataCompat.METADATA_KEY_DURATION)
            counter.duration = duration
            val durationString = formatTime(duration)
            updateDurationText(durationString)
        } else if (data is PlaybackStateCompat) {
            counter.updateState(data)
        }
    }

    private fun updateDurationText(duration: String) {
        this.duration!!.text = duration
    }
}